### PR TITLE
Update OnRCStatus with a new allowed parameter

### PR DESCRIPTION
**Pull request by LitvinenkoIra**
### This implementation is based on OnRCStatus notification, so this PR should be merged after https://github.com/smartdevicelink/sdl_core/pull/2209

Technical task : #2244

**This PR is just for review. Original PR #2306**

This pull request contains the implementation of Remote Control - Update OnRCStatus with a new allowed parameter
(proposal: https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0172-onRcStatus-allowed.md)

Test scripts: https://github.com/smartdevicelink/sdl_atf_test_scripts/pull/1952
For testing use the following test set:
https://github.com/smartdevicelink/sdl_atf_test_scripts/blob/feature/onRcStatus_allowed/test_sets/rc_OnRCStatus.txt

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/resource_allocation_manager_impl.cc [hunk 1]
```diff
@@ -292,17 +300,40 @@ void ResourceAllocationManagerImpl::SetResourceAquired(
...
-    rpc_service_.SendMessageToHMI(msg_to_hmi);
+    smart_objects::SmartObjectSPtr msg_to_mobile;
+    smart_objects::SmartObjectSPtr msg_to_hmi;
+    switch (event) {
+      case NotificationTrigger::APP_REGISTRATION:
+        msg_to_mobile = CreateOnRCStatusNotificationToMobile(rc_app);
+        (*msg_to_mobile)[application_manager::strings::msg_params]
+                        [message_params::kAllowed] = is_rc_enabled();
+        rpc_service_.SendMessageToMobile(msg_to_mobile);
+        if (is_rc_enabled()) {
```
Comment (yang1070) [inline src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/resource_allocation_manager_impl.cc]: why send notification to HMI when an app registers?

Comment (yang1070) [reply src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/resource_allocation_manager_impl.cc]: never mind, now I understand it is the first notification tell the HMI that the app has no allocated modules.

Comment (yang1070) [reply src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/resource_allocation_manager_impl.cc]: notification shall be sent to newly registered RC app, not to existing RC apps, as there is no change to the two lists

**Comment by LitvinenkoIra** [comment]
@yang1070 please look at ffdf19431ba6eb6ccd8bfb6d3cec5e62a3aa4c73
now we send OnRCStatus in following cases:    

- When we register RC app and RC disabled on HMI:
   _SDL->mob (only to mob and  to the newly registered RC app ) OnRCStatus with allowed = false_   
 - When we register RC app and RC enabled on HMI:
   _SDL->mob  (only to mob  to the newly registered RC app) OnRCStatus with allowed = true_    
- When we enable/disable RC:
   _SDL->mob (to all RC apps) OnRCStatus with allowed = true/false
   SDL->HMI OnRCStatus (also for all RC apps) (allocatedModules=[], freeModules=[x,y,z])_    
- Module allocation:
   _SDL->mob OnRCStatus without allowed param
   SDL->HMI OnRCStatus_

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/include/rc_rpc_plugin/resource_allocation_manager.h [hunk 2]
```diff
@@ -160,9 +160,14 @@ class ResourceAllocationManager {
   /**
    * @brief Create and send OnRCStatusNotification to mobile and HMI
    * @param event trigger for notification sending
+   * @param application - application to which notification is sent
```
Comment (yang1070) [inline src/components/application_manager/rpc_plugins/rc_rpc_plugin/include/rc_rpc_plugin/resource_allocation_manager.h]: could you please rephrase this, I'm not quite understand

Comment (LitvinenkoIra) [reply src/components/application_manager/rpc_plugins/rc_rpc_plugin/include/rc_rpc_plugin/resource_allocation_manager.h]: @yang1070 please look at 63375e52c51679b0bbf44e9748a9bfff25a3564e

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/commands/hmi/rc_on_remote_control_settings_notification.cc [hunk 3]
```diff
@@ -150,8 +150,8 @@ void RCOnRemoteControlSettingsNotification::Run() {
   } else {
     LOG4CXX_DEBUG(logger_, "Disallowing RC Functionality");
     DisallowRCFunctionality();
-    resource_allocation_manager_.set_rc_enabled(false);
     resource_allocation_manager_.ResetAllAllocations();
+    resource_allocation_manager_.set_rc_enabled(false);
```
[the remark this answers is not in the document]
Comment (yang1070) [reply src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/commands/hmi/rc_on_remote_control_settings_notification.cc]: I checked ResetAllAllocations, it does not trigger a nonRCStatus notification. That looks good.

File: src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/resource_allocation_manager_impl.cc [hunk 4]
```diff
@@ -174,7 +174,9 @@ void ResourceAllocationManagerImpl::ProcessApplicationPolicyUpdate() {
       if (rc_extention) {
         rc_extention->UnsubscribeFromInteriorVehicleData(*module);
       }
-      SendOnRCStatusNotification();
+      SendOnRCStatusNotifications(
```
Comment (yang1070) [inline src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/resource_allocation_manager_impl.cc]: this is inside a `for` loop. My concern is if there is a policy update that disallows 2 or more moduleTypes from an application, will this trigger 2 or more notifications to the changing app and HMI. It is preferred to trigger one notification to the app/hmi per one policy update no matter how many moduletTypes are disallowed.

Comment (yang1070) [reply src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/resource_allocation_manager_impl.cc]: I'd like to move SendOnRCStatusNotifications() out side of the for loop and call it only if disallowed_modules.size()>0

Comment (LitvinenkoIra) [reply src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/resource_allocation_manager_impl.cc]: @yang1070 Fixed 289cd19

**Comment by yang1070** [comment]
@theresalech Ford has reviewed, tested and approve this PR.

**Pull request closed without merging**